### DISCUSSION TITLE
Vaultwarden homeassistant admin token change not possible

### DISCUSSION THREAD

**Discussion by Stephan-4711** [dani-garcia/vaultwarden#3974] [Q&A]
Hi,

I'm trying to install the vaultwarden container as homeassistant addon.
I'm able to login the admin page.
then I want to change the admin key. 
I enter:
$argon2id$v=.........
The page doesn't complain about the format, when I do a mistake on purpose the admin page gives an error.
After entering the argon2 token I'm not able to login anymore .

I have no clue what I am doing wrong. Is there another way to create the config file inside of the docker container?

**Comment by stefan0xC** [comment]
I'm not familiar with "homeassistant" so can't help you with that but I'd recommend reading the wiki, e.g. [how to enable the `/admin` page](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page) and maybe the [configuration overview](https://github.com/dani-garcia/vaultwarden/wiki/Configuration-overview).

If you have saved the hashed argon2id PHC hash as `ADMIN_TOKEN` successfully, the logs should reflect that. I.e. the following disclaimer will not be displayed anymore:

```
[NOTICE] You are using a plain text `ADMIN_TOKEN` which is insecure.
Please generate a secure Argon2 PHC string by using `vaultwarden hash` or `argon2`.
See: https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#secure-the-admin_token
```

And you should be able to login using the password you have used to create the PHC string.  If you are trying to login using the argon2id PHC string itself this will not work. And it also won't work if you have mistyped your password. (Did you use `vaultwarden hash` to generate the PHC string or another method?)

**Reply by Stephan-4711** [reply]
Finally I was able to find the config file inside the docker container.
I stored the key using the admin page. But it was stored missformated in the config file. Manually adding the "" at the correct position solved my problem

**Reply by MarioShebib** [reply]
@Stephan-4711 which config file in the docker container is it found in since I'm in the same situation of wanting to find the config file and the admin token key of Home Assistant OS but am having trouble trying to find it.

**Reply by jvince11** [reply]
**EDIT:** I got it figured out finally. I think the biggest thing to realize with the admin panel and how it relates to the environment variable "admin_token" in config.json is that `config.json` will add {""} around the entire string entered in the field on the `/admin` panel. So since the hashed string was delivered with the {''} so I believe just deleting each `'` on the ends of the string should solve the mystery. 

Moreover, I couldn't find the `config.json` file at first but I must have only been looking for it before making any changes to the `/admin` panel. My persistent, limited knowledge of how these things work can be the point of blame here. So when I just hit save and was again locked out of the `/admin` panel, but then I noticed in the add-on logs in the home assistant UI that the 

> config.json file was created after saving the wrong format and then restarting the add-on from home assistant. 

! For those who may be in a similar quandary *as I've seen many unfinished threads across the inter webs about this* - you can open the advanced ssh terminal in home assistant and run the command `docker exec -it [Vaultwarden-containerID] /bin/bash` and then I installed a text editor inside the container with `apt install nano -y` - found the `admin_token` line and saw way too many single and double quotes, some of which also were then paired up with escapes `/` in places they didn't need to be. 

I deleted the escapes `/` and any quotation marks except for those that wrapped the entire string after `"admin_token":` and saved the file, exited the container, and restarted the add-on from HA. After that I was able to use the password I provided during the hash creation to re-authenticate with the new admin token and voila, Good to go!

---

> Finally I was able to find the config file inside the docker container.
> 
> I stored the key using the admin page. But it was stored missformated in the config file. Manually adding the "" at the correct position solved my problem 

I've tried a few different ways to enter the full string after "ADMIN_TOKEN" but I keep getting locked out and have to uninstall and reinstall the add-on in HA. Is there a more efficient way to play this for experimenting?

The hash generates a string for me that looks like `ADMIN_TOKEN='$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]'`

Where is proper placement for the additional {""} you mentioned? 
- I wouldn't have guessed that adding {""}  would be a solution since the full phrase seems to be wrapped in {''} already, which I've already tried including in the [ADMIN_TOKEN] field on the `/admin` page. 

Also, does adding the {""} wrapper mean that I would take away the current {''} wrapper?

E.g. 

from:
`ADMIN_TOKEN='$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]'`

To: `ADMIN_TOKEN="$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]"`

**Reply by lordtrader2020** [reply]
hi wher is the location of  config.json   file ??

**Reply by jvince11** [reply]
1) For me, I first had to add the incorrect string in the admin token field on the admin console page in my web browser, this locked me out of the admin page, but I regained access after completing the steps below. 
2) ssh into home assistant, or get the “Advanced terminal” add-on 
3) run command `docker ps | grep bitwarden` and copy the container id for bitwarden. 
    - the container ID might look like this: ** addon_<random alphanumeric string>_bitwarden**
3) paste the copied container id into command: `docker exec -it [containerID] /bin/bash`
4) once you are in the shell for the bitwarden container, run command `ls  data` and you should see the `config.json` file there. 
    - *note that when I entered the container’s shell it was already in the root `/` filesystem. if `ls data` doesn’t work, then try `ls /data`*
5) now just edit the file with whichever editor you like (I needed to download the `nano` itext editor nside the container first. then was able to run command `nano data/config.json`
    - just ensure the  line looks like the following: `"admin_token": "$argon2id…remaining hash value…”`


How that works for you. best of luck!

**Reply by oleksiwd** [reply]
> **EDIT:** I got it figured out finally. I think the biggest thing to realize with the admin panel and how it relates to the environment variable "admin_token" in config.json is that `config.json` will add {""} around the entire string entered in the field on the `/admin` panel. So since the hashed string was delivered with the {''} so I believe just deleting each `'` on the ends of the string should solve the mystery.
> 
> Moreover, I couldn't find the `config.json` file at first but I must have only been looking for it before making any changes to the `/admin` panel. My persistent, limited knowledge of how these things work can be the point of blame here. So when I just hit save and was again locked out of the `/admin` panel, but then I noticed in the add-on logs in the home assistant UI that the
> 
> > config.json file was created after saving the wrong format and then restarting the add-on from home assistant.
> 
> ! For those who may be in a similar quandary _as I've seen many unfinished threads across the inter webs about this_ - you can open the advanced ssh terminal in home assistant and run the command `docker exec -it [Vaultwarden-containerID] /bin/bash` and then I installed a text editor inside the container with `apt install nano -y` - found the `admin_token` line and saw way too many single and double quotes, some of which also were then paired up with escapes `/` in places they didn't need to be.
> 
> I deleted the escapes `/` and any quotation marks except for those that wrapped the entire string after `"admin_token":` and saved the file, exited the container, and restarted the add-on from HA. After that I was able to use the password I provided during the hash creation to re-authenticate with the new admin token and voila, Good to go!
> 
> > Finally I was able to find the config file inside the docker container.
> > I stored the key using the admin page. But it was stored missformated in the config file. Manually adding the "" at the correct position solved my problem
> 
> I've tried a few different ways to enter the full string after "ADMIN_TOKEN" but I keep getting locked out and have to uninstall and reinstall the add-on in HA. Is there a more efficient way to play this for experimenting?
> 
> The hash generates a string for me that looks like `ADMIN_TOKEN='$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]'`
> 
> Where is proper placement for the additional {""} you mentioned?
> 
>     * I wouldn't have guessed that adding {""}  would be a solution since the full phrase seems to be wrapped in {''} already, which I've already tried including in the [ADMIN_TOKEN] field on the `/admin` page.
> 
> 
> Also, does adding the {""} wrapper mean that I would take away the current {''} wrapper?
> 
> E.g.
> 
> from: `ADMIN_TOKEN='$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]'`
> 
> To: `ADMIN_TOKEN="$argon2id$v=[xxx]$m=[xxxxxx],t=[x],p=[string of 89 characters]"`

Perfect.  This helped me with a CasaOS install, also.  Thank you!!!
